### PR TITLE
[DNM] Zero Alloc Zero Copy Key Infra [Part 1]

### PR DESCRIPTION
**Pull request by breeswish**
###  What have you changed?

This PR provides a new key implementation that enables zero allocation and zero copy for higher levels, by exposing the concept of "Physical" key and "Logical" key.

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

### CODE REVIEW

File: components/engine/src/peekable.rs [hunk 1]
```diff
@@ -2,13 +2,27 @@
...
 // TODO: refactor this trait into rocksdb trait.
 pub trait Peekable {
-    fn get_value(&self, key: &[u8]) -> Result<Option<DBVector>>;
-    fn get_value_cf(&self, cf: &str, key: &[u8]) -> Result<Option<DBVector>>;
+    type Key: PhysicalKey;
 
-    fn get_msg<M: protobuf::Message + Default>(&self, key: &[u8]) -> Result<Option<M>> {
+    fn get_value(
+        &self,
+        key: impl ToPhysicalKeySlice<<Self::Key as PhysicalKey>::Slice>,
```
Comment (breeswish) [inline components/engine/src/peekable.rs]: This syntax is ver verbose. Is there any chance we can simplify it, like use `Self::Key::Slice`? (currently it is not supported..) @kennytm @nrc @brson

Comment (kennytm) [reply components/engine/src/peekable.rs]: @breeswish This is rust-lang/rust#38078. I don't think this is gonna be solved soon. If you don't like the FQN here you could define a type alias elsewhere

```rust
type SliceOf<S> = <S::Key as PhysicalKey>::Slice;

...

key: impl ToPhysicalKeySlice<SliceOf<Self>>,
```

**Pull request closed without merging**